presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcFileWriter.java

```diff
@@ -191,11 +249,77 @@ public long getValidationCpuNanos()
         return validationCpuNanos;
     }
 
+    public int getStripeRowCount()
+    {
+        return orcWriter.getStripeRowCount();
+    }
+
+    public void setMaxWriteId(long maxWriteId)
+    {
+        this.maxWriteId = OptionalLong.of(maxWriteId);
+    }
+
+    public OptionalLong getMaxWriteId()
+    {
+        return maxWriteId;
+    }
+
+    public void updateUserMetadata(Map<String, String> userMetadata)
+    {
+        orcWriter.updateUserMetadata(userMetadata);
+    }
+
     @Override
     public String toString()
     {
         return toStringHelper(this)
                 .add("writer", orcWriter)
                 .toString();
     }
+
+    private Block[] buildAcidColumns(Block rowBlock, AcidTransaction transaction)
+    {
+        int positionCount = rowBlock.getPositionCount();
+        int bucketValue = computeBucketValue(bucketNumber.isPresent() ? bucketNumber.getAsInt() : 0, 0);
+        // operation, originalWriteId, bucket, rowId, currentWriteId, row<>
+        return new Block[] {
+                RunLengthEncodedBlock.create(INTEGER, (long) getOrcOperation(transaction), positionCount),
+                RunLengthEncodedBlock.create(BIGINT, transaction.getWriteId(), positionCount),
+                RunLengthEncodedBlock.create(INTEGER, (long) bucketValue, positionCount),
+                buildAcidRowIdsColumn(positionCount),
+                RunLengthEncodedBlock.create(BIGINT, transaction.getWriteId(), positionCount),
+                rowBlock
+        };
+    }
+
+    private int getOrcOperation(AcidTransaction transaction)
+    {
+        switch (transaction.getOperation()) {
+            case INSERT:
+                return 0;
+            default:
+                throw new IllegalStateException("In getOrcOperation, transaction isn't INSERT, transaction " + transaction);
+        }
+    }
+
+    private Block buildAcidRowIdsColumn(int positionCount)
+    {
+        long[] rowIds = new long[positionCount];
+        for (int i = 0; i < positionCount; i++) {
+            rowIds[i] = i;
+        }
+        return new LongArrayBlock(positionCount, Optional.empty(), rowIds);
+    }
+
+    public static int extractBucketNumber(int bucketValue)
+    {
+        return (bucketValue >> 16) & 0xFFF;
+    }
+
+    public static int computeBucketValue(int bucketId, int statementId)
+    {
+        checkArgument(statementId >= 0 && statementId < 1 << 16, "statementId should be non-negative and less than 1 << 16, but is " + statementId);
```

Comment:
Use `%s` so that we don't compute the message on every call

Reply:
Changed.